File: Mice/Surface.m
//
//  Quad.m
//  MeltedIce
//
//  Created by Raphaël Calabro on 24/03/2016.
//  Copyright © 2016 Raphaël Calabro. All rights reserved.
//

#import "Surface.h"

#if TARGET_OS_IOS
#import "MeltedIce-Swift.h"
#else
#import "MeltedIce_OSX-Swift.h"
#endif

@interface Surface () {
    NSUInteger _cursor;
}
@end

@implementation Surface

- (id)initWithMemory:(GLfloat *)memory reference:(NSInteger)reference coordinates:(NSInteger)coordinates {
    self = [super init];
    if (self) {
        _memory = memory;
        _cursor = reference * coordinates * [Surfaces vertexesByQuad];
    }
    return self;
}

- (void)clear {
    memset(_memory + _cursor, 0, [Surfaces vertexesByQuad] * [Surfaces coordinatesByVertice] * sizeof(GLfloat));
}

- (void)setQuadWithLeft:(GLfloat)left top:(GLfloat)top width:(GLfloat)width height:(GLfloat)height {
    [self setQuadWithLeft:left right:left + width top:top bottom:top + height];
}

- (void)setQuadWithSquare:(Square *)square {
    [self setQuadWithLeft:square.left right:square.right top:-square.top bottom:-square.bottom];
}

- (void)setQuadWithLeft:(const GLfloat)left right:(const GLfloat)right top:(const GLfloat)top bottom:(const GLfloat)bottom {
    // Bas gauche
    _memory[_cursor] = left;
    _memory[_cursor + 1] = bottom;
    
    // (idem)
    _memory[_cursor + 2] = left;
    _memory[_cursor + 3] = bottom;
    
    // Bas droite
    _memory[_cursor + 4] = right;
    _memory[_cursor + 5] = bottom;
    
    // Haut gauche
    _memory[_cursor + 6] = left;
    _memory[_cursor + 7] = top;
    
    // Haut droite
    _memory[_cursor + 8] = right;
    _memory[_cursor + 9] = top;
    
    // (idem)
    _memory[_cursor + 10] = right;
    _memory[_cursor + 11] = top;
}

- (void)setQuadWithQuadrilateral:(Quadrilateral *)quadrilateral {
    // bas gauche
    _memory[_cursor] = quadrilateral.bottomLeft.x;
    _memory[_cursor + 1] = -quadrilateral.bottomLeft.y;
    
    // (idem)
    _memory[_cursor + 2] = quadrilateral.bottomLeft.x;
    _memory[_cursor + 3] = -quadrilateral.bottomLeft.y;
    
    // bas droite
    _memory[_cursor + 4] = quadrilateral.bottomRight.x;
    _memory[_cursor + 5] = -quadrilateral.bottomRight.y;
    
    // haut gauche
    _memory[_cursor + 6] = quadrilateral.topLeft.x;
    _memory[_cursor + 7] = -quadrilateral.topLeft.y;
    
    // haut droite
    _memory[_cursor + 8] = quadrilateral.topRight.x;
    _memory[_cursor + 9] = -quadrilateral.topRight.y;
    
    // (idem)
    _memory[_cursor + 10] = quadrilateral.topRight.x;
    _memory[_cursor + 11] = -quadrilateral.topRight.y;
}

- (void)setColor:(Color *)color {
    const NSInteger vertexesByQuad = [Surfaces vertexesByQuad];
    NSInteger index = _cursor;
    for (NSInteger vertex = 0; vertex < vertexesByQuad; vertex++) {
        _memory[index++] = color.red;
        _memory[index++] = color.green;
        _memory[index++] = color.blue;
        _memory[index++] = color.alpha;
    }
}

- (void)setColorWithWhite:(GLfloat)white alpha:(GLfloat)alpha {
    const NSInteger vertexesByQuad = [Surfaces vertexesByQuad];
    NSInteger index = _cursor;
    for (NSInteger vertex = 0; vertex < vertexesByQuad; vertex++) {
        _memory[index++] = white;
        _memory[index++] = white;
        _memory[index++] = white;
        _memory[index++] = alpha;
    }
}

@end

File: Melisse/Surface.swift
//
//  Surface.swift
//  Melisse
//
//  Created by Raphaël Calabro on 04/05/2016.
//  Copyright © 2016 Raphaël Calabro. All rights reserved.
//

import GLKit

public class Surface<Element : Numeric> {
    
    let memory: UnsafeMutablePointer<Element>
    let coordinates: Int
    let vertexesByQuad: Int
    
    init(memory: UnsafeMutablePointer<Element>, coordinates: Int, vertexesByQuad: Int) {
        self.memory = memory
        self.coordinates = coordinates
        self.vertexesByQuad = vertexesByQuad
    }
    
    public func clear() {
        memset(memory, 0, vertexesByQuad * coordinates * MemoryLayout<Element>.size)
    }
    
    public func setQuadWith(left: Element, top: Element, width: Element, height: Element) {
        setQuadWith(left: left, right: left + width, top: top, bottom: top + height)
    }
    
    public func setQuadWith(left: Element, right: Element, top: Element, bottom: Element) {
        // Bas gauche
        memory[0] = left
        memory[1] = bottom
        
        // (idem)
        memory[2] = left
        memory[3] = bottom
        
        // Bas droite
        memory[4] = right
        memory[5] = bottom
        
        // Haut gauche
        memory[6] = left
        memory[7] = top
        
        // Haut droite
        memory[8] = right
        memory[9] = top
        
        // (idem)
        memory[10] = right
        memory[11] = top
    }
    
    public func setColor(_ color: Color<Element>) {
        var index = 0
        for _ in 0 ..< vertexesByQuad {
            memory[index] = color.red
            memory[index + 1] = color.green
            memory[index + 2] = color.blue
            memory[index + 3] = color.alpha
            index += 4
        }
    }
    
    public func setColor(with white: Element, alpha: Element) {
        var index = 0
        for _ in 0 ..< vertexesByQuad {
            memory[index] = white
            memory[index + 1] = white
            memory[index + 2] = white
            memory[index + 3] = alpha
            index += 4
        }
    }
    
    public func setAlpha(_ alpha: Element) {
        var index = 0
        for _ in 0 ..< vertexesByQuad {
            memory[index + 3] = alpha
            index += 4
        }
    }
    
    public func setGradient(_ gradient: Gradient<Element>) {
        // Bas gauche
        memory[0] = gradient.bottomLeft.red
        memory[1] = gradient.bottomLeft.green
        memory[2] = gradient.bottomLeft.blue
        memory[3] = gradient.bottomLeft.alpha
        
        // (idem)
        memory[4] = gradient.bottomLeft.red
        memory[5] = gradient.bottomLeft.green
        memory[6] = gradient.bottomLeft.blue
        memory[7] = gradient.bottomLeft.alpha
        
        // Bas droite
        memory[8] = gradient.bottomRight.red
        memory[9] = gradient.bottomRight.green
        memory[10] = gradient.bottomRight.blue
        memory[11] = gradient.bottomRight.alpha
        
        // Haut gauche
        memory[12] = gradient.topLeft.red
        memory[13] = gradient.topLeft.green
        memory[14] = gradient.topLeft.blue
        memory[15] = gradient.topLeft.alpha
        
        // Haut droite
        memory[16] = gradient.topRight.red
        memory[17] = gradient.topRight.green
        memory[18] = gradient.topRight.blue
        memory[19] = gradient.topRight.alpha
        
        // (idem)
        memory[20] = gradient.topRight.red
        memory[21] = gradient.topRight.green
        memory[22] = gradient.topRight.blue
        memory[23] = gradient.topRight.alpha
    }
    
}

public extension Surface where Element: FloatingPoint, Element: Signed {
    
    func setQuadWith(left: Int, top: Int, width: Int, height: Int, direction: Direction, texture: GLKTextureInfo) {
        setQuadWith(left: (Element(left) + Element(width) * direction.mirror) / Element(texture.width),
                    top: Element(top) / Element(texture.height),
                    width: (Element(width) * direction.value) / Element(texture.width),
                    height: Element(height) / Element(texture.height))
    }
    
    func setQuadWith(_ rectangle: Rectangle<Element>) {
        setQuadWith(left: rectangle.left, right: rectangle.right, top: -rectangle.top, bottom: -rectangle.bottom)
    }
    
    func setQuadWith(_ quadrilateral: Quadrilateral<Element>) {
        memory[0] = quadrilateral.bottomLeft.x
        memory[1] = -quadrilateral.bottomLeft.y
        
        // (idem)
        memory[2] = quadrilateral.bottomLeft.x
        memory[3] = -quadrilateral.bottomLeft.y
        
        // bas droite
        memory[4] = quadrilateral.bottomRight.x
        memory[5] = -quadrilateral.bottomRight.y
        
        // haut gauche
        memory[6] = quadrilateral.topLeft.x
        memory[7] = -quadrilateral.topLeft.y
        
        // haut droite
        memory[8] = quadrilateral.topRight.x
        memory[9] = -quadrilateral.topRight.y
        
        // (idem)
        memory[10] = quadrilateral.topRight.x
        memory[11] = -quadrilateral.topRight.y
    }
    
}

public extension Surface where Element: Integer {
    
    func setQuadWith(left: Element, top: Element, width: Element, height: Element, direction: Direction, texture: GLKTextureInfo) {
        setQuadWith(left: left + width * Element(direction.mirror),
                    top: top,
                    width: width * Element(direction.value),
                    height: height)
    }
    
}
